src/container/sycl_iterator.hpp

```diff
@@ -90,6 +93,13 @@ BufferIterator<element_t, codeplay_policy>::operator++(int i) {
   return temp_iterator;
 }
 
+// implicitly converting non-const buffer to const buffer
+template <typename element_t>
+inline BufferIterator<element_t, codeplay_policy>::
+operator BufferIterator<const element_t, codeplay_policy>() const {
+  return BufferIterator<const element_t, codeplay_policy>(*this);
+}
```

Comment:
I think this should be a constructor instead of a conversion operator. You may need to SFINAE out the case when `element_t` is `const` to avoid overload resolution ambiguity.